unused_must_use: Don't warn on `Result<(), Uninhabited>` or `ControlFlow<Uninhabited, ()>`

joshtriplett opened the pull request:
This suppresses warnings on things like `Result<(), !>`, which helps simplify code using the common pattern of having an `Error` associated type: code will only have to check the error if there is a possibility of error.

This will, for instance, help with future refactorings of `write!` in the standard library.

As this is a user-visible change to lint behavior, it'll require a lang FCP.

---

My proposal, here, is that we handle this simple case to make common patterns more usable. This does not rule out the possibility of adding more cases in the future, including general trait-based cases. However, I don't think we should make this common case wait on the more general cases. In particular, this solution does not close any doors on replacing this special case with a general case.

This would unblock some planned work in the standard library to make `write!` more usable for infallible cases (e.g. writing into a `Vec` or `String`).

joshtriplett committed:
unused_must_use: Factor out a variable for the parent module DefId

This simplifies the initial conditional, and will allow reusing the
variable in subsequent checks.

joshtriplett committed:
unused_must_use: Don't warn on `Result<(), Uninhabited>`

This suppresses warnings on things like `Result<(), !>`, which helps
simplify code using the common pattern of having an `Error` associated
type: code will only have to check the error if there is a possibility
of error.

fmease commented:
I'll be waiting for the outcome of T-lang's triage meeting before reviewing this since the proposed rules might very well change drastically (e.g., trait-driving, not trait-driven, etc.; cc Zulip discussion).

joshtriplett committed:
unused_must_use: Add test for a method using an associated error type

Amanieu commented on compiler/rustc_lint/src/unused.rs:
```diff
@@ -293,6 +291,22 @@ impl<'tcx> LateLintPass<'tcx> for UnusedResults {
+                        )
+                        && !matches!(
+                            args.type_at(1).kind(),
+                            ty::Adt(err_def, _) if err_def.is_variant_list_non_exhaustive()
```
I believe `#[non_exhaustive]` is already handled by `is_inhabited_from`.

Amanieu replied:
That's because `#[non_exhaustive]` is ignored for local types from the current crate. It only applies when using the type from another crate.

I think we should be consistent with the behavior of other parts of the compiler that handle exhaustiveness checking in this regards.

joshtriplett replied:
@Amanieu Yes, I did realize that; I had used `is_variant_list_non_exhaustive` intentionally in order to make it treat types from the *same* crate as inhabited if `non_exhaustive`. But you're right that if not trying to achieve that, I could leave it out and get the "`non_exhaustive` from a *different* crate" behavior for free.

Amanieu commented:
As a general rule I think this should follow the same rules what `let Ok(()) = foo` would accept.

joshtriplett committed:
unused_must_use: Suppress warnings for `ControlFlow<Uninhabited, ()>` too

niacdoial commented:
huh...
github doesn't let me choose your branch as a PR destination.
er... anyway, I would like to contribute a test (making sure unresolved type aliases behave like inhabited types), if you'll let me:
https://github.com/niacdoial/rust-contrib-fork/tree/must_use_uninhab

RalfJung commented on tests/ui/lint/unused/must_use-result-unit-uninhabited.rs:
```diff
@@ -0,0 +1,93 @@
+}
+
+fn main() {
+    f1(); //~ ERROR: unused `Result` that must be used
+    f2();
+    f3();
+    f4();
+    f5(); //~ ERROR: unused `Result` that must be used
+    f6(S1);
+    f6(S2); //~ ERROR: unused `Result` that must be used
+    S1.method();
+    S2.method(); //~ ERROR: unused `Result` that must be used
+
+    c1(); //~ ERROR: unused `ControlFlow` that must be used
+    c2();
+    c3();
```
It is tricky to review this for correctness since the function names are all meaningless, so one has to cross-reference them all manually. Can you rename the functions to make it more clear which ones should get the lint and which not?

joshtriplett replied:
Done.

joshtriplett committed:
unused_must_use: Rename test functions to reflect what they test

This makes it easier to review without cross-referencing each test
function with its invocation.

niacdoial committed:
unused_must_use: Add extra test for types that are still generic

joshtriplett commented:
@scottmcm I don't want to generalize that here yet, because someone might e.g. have some `#[must_use]` ZST that represents "capture this token and pass it to some other function to prove you got an `Ok`", and I wouldn't want to override that. Some more general design *could* argue for that, but `()` is the case that I'm confident any more general solution will encompass.

fmease reviewed (approved):
r=me once the FCP completes and with minor nits addressed in one way or another

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147382/828c2a9afccf3b3ff8133368cfbc8bfe526aaa4d..f4ca5fbbaea56c85456858c33437629ca55514c9)*

joshtriplett committed:
Factor out a helper function to check if a type is uninhabited

WaffleLapkin commented on compiler/rustc_lint/src/unused.rs:
```diff
@@ -293,6 +293,22 @@ impl<'tcx> LateLintPass<'tcx> for UnusedResults {
+                // Suppress warnings on `Result<(), Uninhabited>` (e.g. `Result<(), !>`).
+                ty::Adt(def, args)
+                    if cx.tcx.is_diagnostic_item(sym::Result, def.did())
+                        && args.type_at(0).is_unit()
```
Is there a reason why this checks for `T = ()`? `Result<T, !>` is _basically_ `T`, so I'd expect this to not check for `()`. In #147854, which I opened because I haven't seen this PR, I just return `is_ty_must_use(cx, args.type_at(0), expr, span)` here.

joshtriplett replied:
@WaffleLapkin I can see the argument for that, but it'd be less clear-cut. `Result<(), !>` has only one possible value, so there's zero value in checking it at all, and it seems safe to unconditionally make it ignore `must_use`. Something like `Result<u32, !>`, on the other hand, would *currently* be flagged by the `must_use` on `Result`, and it would be a potential surprise to people to stop doing that; people might be expecting to have to capture the return value with `let Ok(value) = ...`, for instance.

If we were going to potentially suppress warnings on `Result<T, !>`, I think that'd need a separate FCP. We might also want to let people know that they may want to mark the *function* `#[must_use]` if they were previously counting on the `Result` to do that.

Separately from that, if we're checking for `T` being `#[must_use]`, it'd be nice to have a design where a function returning `Result<T, E>` produces an `unused_must_use` warning if you apply a `?` and then ignore the `T`. But that doesn't have to happen right away, and I'm all for incremental results.

One way or another, though, I think that would be an additional change requiring an additional FCP.

bors commented:
:pushpin: Commit 7aa7ecc98234144dcdde6e2e9326ac806c0f36c3 has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 7aa7ecc98234144dcdde6e2e9326ac806c0f36c3 -->
<!-- homu: {"type":"Approved","sha":"7aa7ecc98234144dcdde6e2e9326ac806c0f36c3","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
